keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,604 @@
+
+## Motivation
+
+In real life, workloads do vertical scaling until certain points but then start to do partition, at the same time provide a software layer to tweak key parameters like iops/throughput.
+
+Currently after CreateVolume with provider specific parameters pass in storage class, Kubernetes does not handle update of QoS related parameters update. The only way to achieve this is to use the cloud provider's API directly to modify iops and throughput. This creates frictions of managing storage resources on k8s. Allowing Resources to be changed without managing it through different provider's APIs directly simplifies this whole process.
+
+### Goals
+
+* Cloud-provider and storage vendor independent specification of basic QoS parameters
+* Specifying QoS at volume creation
+* Changing QoS parameters on-line for existing volumes
+* Specify quotas in a similar way to that of volume capacity
+
+### Non-Goals
+
+* OS-level IO QoS, for example cgroup policy
+* Inter-pod QoS. This KEP is restricted to what QoS can be defined on a volume.
+* Volume QoS classes. This proposal is for fine-grained provisioned IO control, and not for defining general classes of volume performance. Currently, storage vendors may accomplish this by adding custom parameters to a storage class. A future KEP may codify such behavior with common storage class parameter names, but that is out of scope of this KEP.
+* Scheduling based on total IO limits of a node. For example, total throughput for many cloud provider network attached storage volumes is limited by instance network bandwidth and may prevent achieving the provisioned IO. Scheduling to mitigate these factors is out of scope but it should be considered as a future extension of this KEP.
+
+## Proposal
+
+### Kubernetes API
+Add two well-known ResourceName keys to the existing ResourceList type used in PersistentVolumeClaim.Spec.Resources.{Limits,Requests} and PersistentVolume.Spec.Capacity:
+* (ReadWrite) IOPS independent from capacity
+* (ReadWrite) Throughput
+These values can be changed on a PersistentVolumeClaim, and will be reconciled against the storage provider in the same way as storage capacity resizing. The status of a PVC will reflect the currently provisioned IO as reported by the storage provider (in the same way that the status states the current size of the volume during a resize operation).
+
+```
+// Example:
+kind: PersistentVolumeClaim
+...
+spec:
+  ...
+  resources:
+requests:
+  storage: 4Gi
+  throughput: "1000M" // MiB/s as unit
```

Comment:
What is the desired behavior if a driver does not support throughput and/or iops? Should the fields be ignored or should the PVC not be bound?

Reply:
If the driver does not support iops/throughput, it should return an unretriable error.